consensus/src/pipeline/decryption_pipeline_builder.rs

```diff
@@ -0,0 +1,135 @@
+use aptos_batch_encryption::{schemes::fptx::FPTX, traits::BatchThresholdEncryption};
+use aptos_consensus_types::{
+    block::Block,
+    common::Author,
+    pipelined_block::{DecryptionResult, MaterializeResult, TaskFuture, TaskResult},
+};
+use aptos_types::{
+    secret_sharing::{
+        Ciphertext, DigestKey, MasterSecretKeyShare, SecretShare, SecretShareConfig,
+        SecretShareMetadata, SecretSharedKey,
+    },
+    transaction::encrypted_payload::DecryptedPayload,
+};
+use rayon::iter::{IndexedParallelIterator, IntoParallelIterator, ParallelIterator};
+use std::sync::Arc;
+use tokio::sync::oneshot;
+
+impl PipelineBuilder {
+    /// Precondition: Block is materialized and the transactions are available locally
+    /// What it does: Decrypt encrypted transactions in the block
+    pub(crate) async fn decrypt_encrypted_txns(
+        materialize_fut: TaskFuture<MaterializeResult>,
+        block: Arc<Block>,
+        author: Author,
+        secret_share_config: Option<SecretShareConfig>,
+        derived_self_key_share_tx: oneshot::Sender<Option<SecretShare>>,
+        secret_shared_key_rx: oneshot::Receiver<Option<SecretSharedKey>>,
+    ) -> TaskResult<DecryptionResult> {
+        let mut tracker = Tracker::start_waiting("decrypt_encrypted_txns", &block);
+        let (input_txns, max_txns_from_block_to_execute, block_gas_limit) = materialize_fut.await?;
+
+        tracker.start_working();
+
+        if secret_share_config.is_none() {
+            return Ok((input_txns, max_txns_from_block_to_execute, block_gas_limit));
+        }
+
+        let (encrypted_txns, unencrypted_txns): (Vec<_>, Vec<_>) = input_txns
+            .into_iter()
+            .partition(|txn| txn.is_encrypted_txn());
+
+        if encrypted_txns.is_empty() {
+            return Ok((
+                unencrypted_txns,
+                max_txns_from_block_to_execute,
+                block_gas_limit,
+            ));
+        }
+
+        let digest_key: DigestKey = secret_share_config.as_ref().unwrap().digest_key().clone();
+        let msk_share: MasterSecretKeyShare =
+            secret_share_config.as_ref().unwrap().msk_share().clone();
+
+        // TODO(ibalajiarun): FIXME
+        let len = 10;
+        let encrypted_txns = if encrypted_txns.len() > len {
+            let mut to_truncate = encrypted_txns;
+            to_truncate.truncate(len);
+            to_truncate
+        } else {
+            encrypted_txns
+        };
+
+        let txn_ciphertexts: Vec<Ciphertext> = encrypted_txns
+            .iter()
+            .map(|txn| {
+                // TODO(ibalajiarun): Avoid clone and use reference instead
+                txn.payload()
+                    .as_encrypted_payload()
+                    .expect("must be a encrypted txn")
+                    .ciphertext()
+                    .clone()
+            })
+            .collect();
+
+        let encryption_round = block.round();
```

Comment:
There is one optimization to save the number of trusted setup: the encryption_round can be decoupled from block round, and increment only when a block contains encrypted txns. Then the trusted setup size is number of encrypted txn blocks instead of blocks. Not sure how important it is for now, but just a note for potential optimizations.

Reply:
interesting, I will add this as TODO. We can revisit this when Rex is back.